src/libtls/tls_server.c

```diff
@@ -189,24 +192,51 @@ static bool select_suite_and_key(private_tls_server_t *this,
 			 this->server);
 		return FALSE;
 	}
+	type = key->get_type(key);
 
-	if (this->tls->get_version_max(this->tls) >= TLS_1_3)
+	if (version_max >= TLS_1_3)
 	{
-		/* currently no support to derive key based on client supported
-		 * signature schemes */
+		array_t *types;
+		enumerator_t *enumerator;
+		key_type_t *element;
+		chunk_t hashsig;
+		bool found = FALSE;
+
 		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
 												  		count, KEY_ANY);
 		if (!this->suite)
 		{
 			DBG1(DBG_TLS, "received cipher suites unacceptable");
 			return FALSE;
 		}
+
+		hashsig = this->hashsig_cert.len ? this->hashsig_cert : this->hashsig;
+		types = tls_create_common_key_type(hashsig, version_min, version_max);
+		enumerator = array_create_enumerator(types);
+		while (enumerator->enumerate(enumerator, &element))
+		{
+			if (*element == type)
+			{
+				found = TRUE;
+				break;
+			}
+			key->destroy(key);
+			key = lib->credmgr->get_private(lib->credmgr, type, this->server,
+											this->server_auth);
+			type = key->get_type(key);
+		}
+		enumerator->destroy(enumerator);
+		array_destroy(types);
+		if (!found)
+		{
+			DBG1(DBG_TLS, "no client supporting key found");
+			return FALSE;
+		}
 	}
 	else
 	{
 		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
-												  		count,
-												  		key->get_type(key));
+												  		count, type);
```

Comment:
I think the above approach could also be used in legacy TLS condition branch, although the logic is a bit different. Also I think I would have to consider EdDSA of [RFC 8422](https://tools.ietf.org/html/rfc8422) you mentioned in #16, right?